### DISCUSSION TITLE
KM271-WiFi mit dewennis FW verliert Verbindung zu Home Assistant

### DISCUSSION THREAD

**Discussion by paxrsi** [dewenni/ESP_Buderus_KM271#108] [Q&A] [answered]
Hallo alle,

Ich habe ein KM271-WiFi mit der dewennis FW Version 4.1.0. Im Prinzip funktioniert die Anbindung an Home Assistant einwandfrei. Aber es gibt zwei Probleme:

1.) Beim Durchstarten von Home Assistant geht die Verbindung reproduzierbar verloren. D.h. ich starte Home Assistant durch (z.B. weil die HA-VM wegen eines Backups in Proxmox gestoppt und nach dem Backup wieder gestartet wird, aber auch wenn ich in Home Assistant ein Update einspiele, das HA neu startet) und danach werden keine Daten mehr vom KM271 aufgezeichnet.
Wenn ich dann im Webinterface des KM271 auf Einstellungen > Neustart gehe, funktioniert die Aufzeichnung nach dem Neustart wieder - bis zum nächsten Durchstarten von HA.
Hat jemand eine Idee, was man da tun kann?

2.) Das Home Assistant Logfile hat einige Errors / Warnings, von denen leider nicht erkenntlich ist, ob die auf das Modul zurückzuführen sind. Ich stelle sie hier einfach mal ein. Vielleicht hat hier jemand eine Idee. Zumindest die ersten Fehler tauchen im Log unmittelbar nach  Logamatic-Einträgen auf.

```
2024-07-21 10:04:45.634 INFO (MainThread) [homeassistant.components.mqtt.discovery] Found new component: sensor Logamatic flash
2024-07-21 10:04:45.745 ERROR (MainThread) [homeassistant.helpers.template] Template variable error: 'value_json' is undefined when rendering '{{ value_json.uptime }}'
2024-07-21 10:04:45.756 ERROR (MainThread) [homeassistant.helpers.template] Template variable error: 'value_json' is undefined when rendering '{{ value_json.restart_reason }}'
2024-07-21 10:04:45.756 ERROR (MainThread) [homeassistant.helpers.template] Template variable error: 'value_json' is undefined when rendering '{{ value_json.flash.split(' ')[0] }}'
2024-07-21 10:04:45.756 ERROR (MainThread) [homeassistant.helpers.template] Template variable error: 'value_json' is undefined when rendering '{{ value_json.heap.split(' ')[0] }}'
```

`2024-07-21 10:06:06.167 WARNING (MainThread) [homeassistant.components.mqtt.climate] Invalid modes mode: manual`

Danke und Grüße
Rainer

**Comment by the78mole** [comment]
Hall Rainer,
Hallo Sven,

Bezüglich des 1. Problems habe ich eine Vermutung. Zwar nutze ich selbst ja meine ESPhome Variante, aber in einem anderen Projekt in dem ich MQTT und HA Discovery verwende, habe ich bemerkt, dass Home Asssitant nach jedem Start eine Discovery-Nachricht erwartet. Ich bin mir jetzt nicht sicher, welcher Teil am Ende die Lösung war, aber ich habe einerseits die Discovery-Nachrichten mit "retain" versehen, andererseits verschicke ich diese nun alle 15 Minuten.

Zu 2 kann ich leider nicht so viel sagen. Bei mir tauchen auch von Zeit zu Zeit mal diese Template-Errors in anderen MQTT-Geräten auf. Du kannst mal mit einem MQTT-Explorer draufschnüffeln und Dir ansehen, was in dem Template-Teil der Discovery und in der eigentlichen Nachricht selbst steht. Aber value_json ist eigentlich der Body der Datennachricht. Und wenn da irgendein Fehler im JSON ist oder gar keine (sinnvollen) Daten drin stecken, wirft er vermutlich genau den Fehler. Ich würde das Log mal beobachten und wenn der Fehler kommt, die letzte Daten-Message durch einen JSON-Parser schieben. Das sollte uns dem Problem etwas näher bringen.

Grüße,
Daniel

**Comment by paxrsi** [comment]
Hallo @dewenni ,
ich wollte mal nachfragen, ob das Problem zu 1.) mit dem neuen Firmwareupdate behoben ist bzw. ob ich noch irgendwie unterstützen kann. Das Problem zu 2. ist nachrangig. Aber 1. ist schon etwas nervig, wenn ich nach jedem Backup das Modul durchstarten muss...

Danke und viele Grüße
Rainer

**Comment by dewenni** [comment]
@paxrsi 
Ich denke es ist mittlerweile klar, warum das Problem mit MQTT Discovery besteht, wenn HA neu startet.

https://www.home-assistant.io/integrations/mqtt/#how-to-use-discovery-messages

Ich werde wohl die Variante umsetzten wo ich auf dem ESP auf die Nachricht `homeassistant/status` reagiere und damit ein erneutes senden der Config Nachrichten triggere.
Sollte das nicht zuverlässig funktionieren, dann sende ich die Config Nachrichten als `retain`. 

Aktuell ist leider beides nicht umgesetzt.
Ich mache die Tage aber ein neues Release wo das dann gefixt seien sollte.

**Comment by dewenni** [comment, marked as answer]
@paxrsi 
Ich habe es im aktuellen Release umgesetzt [v4.1.2](https://github.com/dewenni/ESP_Buderus_KM271/releases/tag/main)
bitte testen.

**Comment by paxrsi** [comment]
@dewenni 
Ich habe heute das Update auf 4.1.2 eingespielt und mehrfach HA durchgestartet, sowohl über HA selbst, als auch über die VM. Die Verbindung zum KM271 war jedes Mal nach dem Neustart sofort und ohne weiteres Zutun wieder da. Bisher ist mir nichts problematisches aufgefallen. Ich würde sagen, das passt so. Wenn ich nochmal was bemerke, dann melde ich mich.

Vielen Dank für Deine Unterstützung!
Rainer

**Reply by dewenni** [reply]
Sehr gut, freut mich!

**Reply by Eule01x** [reply]
Bei mir fehlen nach einem Neustart von HA ein paar Dinge, siehe Screenshot
![grafik](https://github.com/user-attachments/assets/ca1e334e-4b9d-4cee-a5af-87d0e640758a)

Neustart und Deaktivieren/Reaktivieren von MQTT in der FW brachte keine Besserung. Der Rest scheint aber wohl zu funktionieren.

**Reply by Eule01x** [reply]
Ich korrigiere: HK1 Tag-/Nachttemp einstellen über HA / MQTT führte zu keiner Reaktion. Bin auf 4.1.1 zurück, da klappte es wieder. Auch sind zuvor gezeigte Entitäten auch wieder verfügbar.
Vermute, irgendwas ist bei der 4.1.2 durcheinander gekommen

**Reply by dewenni** [reply]
@Eule01x du hast recht - danke für den Hinweis.
Ich habe leider bei der letzten Änderung wo ich noch einen Button für den Restart hinzugefügt habe, einen kleinen aber entscheidenden Fehler eingebaut, so dass die anderen Sollwerte alle nicht mehr funktionierten.
Ich habe das nun im Release v4.1.3 wieder gefixt. Jetzt sollte alles passen.
Sorry!

**Reply by dewenni** [reply]
@Eule01x , @paxrsi 
Das mit die zwei Werte für "HKx_Sommer_ab" und "WW_Zirkukation" nicht funktionieren, liegt daran, dass ich die Konfiguration für die beiden werte geändert habe und HA jetzt aber noch die alte Konfiguration anzeigt, die aber nicht mehr funktioniert.
U das zu beheben muss man eine entsprechende MQTT Nachricht ohne Payload schicken. Damit wird die Konfiguration zurückgesetzt.

Für die beiden Werte wäre das eine leere Nachricht an diese Topics:
 `homeassistant/number/Logamatic/HK1_Sommer_ab/config`  
 `homeassistant/sensor/Logamatic/WW_Zirkulation/config`

alternativ habe ich noch zwei Telnet Befehle ergänzt, mit denen man alle Konfigurationen löschen kann und neu senden kann.

<img width="650" alt="image" src="https://github.com/user-attachments/assets/c19b0f02-72a4-47e9-a871-2f410a33d7a8">

**Comment by paxrsi** [comment]
@dewenni : Danke!

Zur Info:

Ich hatte Schwierigkeiten, die 4.1.3 bei mir einzuspielen. Ich bin wie folgt vorgegangen:
1.) Ich lade die buderus_km271_ota_update_v4.1.3.bin hier von GitHub herunter (1.301.968 Bytes)
2.) Ich gehe in der Weboberfläche auf Tools und wähle dort bei "OTA Firmware Update" bei "Durchsuchen" diese Datei aus.
3.) Dann klicke ich darunter auf "Update"

Je nach Versuch passierte dann folgendes:
4.) Der Fortschrittsbalken wandert und bleibt irgendwann stehen, z.B. bei 30% oder auch bei 45%
5.) Der Fortschrittsbalken wandert bis zum Ende und es erschein ein Dialog mit folgendem Text:
Update nicht erfolgreich!
Ferhler: Could Not Activate The Firmware
Bitte das Update erneut ausführen!

(Nebeninfo: Rechtschreibfehler => Ferhler)

Wenn er bei 4.) irgendwo stehen bleibt und ich dann nochmal die Schritte 2 und 3 ausführe, wandert der Fortschritt ein kleines Stück weiter. Da kam dann immer ein Dialog, dass das Update nicht eingespielt werden konnte (ein anderer als der oben).

Ich hatte auch schon das Modul per "Neu starten" zurückgesetzt, aber das hatte leider auch nicht geholfen. Manchmal kam ich bis zum Ende und dann war das Modul ein paar Minuten down und nicht erreichbar, danach wieder mit der 4.1.2 erreichbar.

Die Logs hatte ich auch aktiviert, aber da wurde nie was angezeigt.

So, und jetzt die Lösung: Ich hatte die buderus_km271_ota_update_v4.1.3.bin auf einem Samba-Share liegen. Zuletzt hatte ich die Datei auf die lokale C:-Platte kopiert - und von da ging es dann. Ob das Zufall war oder einfach nur dieses Mal ein glücklicher Umstand => keine Ahnung. Normaler Weise habe ich mit dem NAS keine Schwierigkeiten. Der Browser ist FireFox unter Windows.

Wenn ich hier schon am Schreiben bin, vielleicht noch ein Wunsch für zukünftige Firmware-Updates: Das Model sendet im Moment alle 10 Sekunden eine Uptime-Message an HomeAssistant. Das führt leider dazu, dass das Historienfenster in HomeAssistant mit sehr vielen Meldungen dieser Art geflutet wird. Das ist einerseits ganz nett, wenn ich sehen will, ob die Verbindung besteht. Andererseits macht es das auch unübersichtlich, wenn ich nach was anderem schauen will. Vielleicht kann man das ja konfigurierbar machen, dass man eine Anzahl an Sekunden für ein Intervall angibt, zu dem diese Uptime-Nachrichten gesendet werden sollen. 0 oder negative Werte könnten dann ausgeschaltet bedeuten - oder man macht noch einen extra Schalter für An / Aus.


Danke und Grüße
Rainer

**Reply by dewenni** [reply]
@paxrsi 

Das mit dem OTA Update ist leider immer so eine Sache. Manchmal klappt das auf Anhieb und manchmal brauche auch ich mehrere Versuche bis es durchläuft. Ich weiß aber ehrlich gesagt nicht woran das liegt. Vermutlich liegt es einfach an der Verbindung und es kommt hier bei den geringsten Störungen zum Fehler. Eine andere Erklärung habe ich leider bisher nicht.
Ich arbeite im Hintergrund noch an einem Update auf Arduino 3 und der Unterstützung für eine Ethernet Schnittstelle.
Mal sehen wie es dann damit läuft.

Das mit dem Update bei HA muss ich mir mal ansehen. Das betrifft mehrere Statusmeldungen die ich zyklisch per MQTT verschicke. Unabhängig ob HA oder nicht. Aber HA wertet diese Nachricht eben auch aus. (Wifi Status und System-Infos)
Problematisch ist aber wohl tatsächlich nur der Wert "uptime" aus der sysInfo Nachricht. Diese Nachricht enthält mehrere Elemente und wird alle 10s geschickt. Ich denke ich schmeiße den Wert `uptime` einfach raus aus der HA Config. Der Status ob online/offline wird eh unabhängig davon gebildet.
